### PR TITLE
Add definitions for 16 packages of the @feathersjs scope

### PR DESCRIPTION
**Pull request by j2L4e**
Please fill in this template.

- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not provide its own types, and you can not add them.

The documentation of feathers' core packages can be found here: https://docs.feathersjs.com/
The respective repos are under the @feathersjs org: https://github.com/feathersjs

### CODE REVIEW

File: types/feathersjs__authentication-client/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+// Project: http://feathersjs.com/
+// Definitions by: Abraao Alves <https://github.com/AbraaoAlves>, Jan Lohage <https://github.com/j2L4e>
+// Definitions: https://github.com/feathersjs-ecosystem/feathers-typescript
+
+import './feathersjs__feathers';
+
+export default function feathersAuthClient(config?: FeathersAuthClientConfig): () => void;
+
+export interface FeathersAuthClientConfig {
+    storage?: Storage;
```
Comment (erikma) [inline types/feathersjs__authentication-client/index.d.ts]: nit: Prefer doc comments here, even if a copy of what's available on other sites, to provide autocomplete self-documentation

File: types/feathersjs__authentication-client/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    cookie?: string;
+    storageKey?: string;
+    jwtStrategy?: string;
+    path?: string;
+    entity?: string;
+    service?: string;
+}
+
+export interface FeathersAuthCredentials {
+    strategy: string;
```
Comment (erikma) [inline types/feathersjs__authentication-client/index.d.ts]: nit: A doc comment with the valid strategy names

File: types/feathersjs__authentication-jwt/feathersjs__authentication-jwt-tests.ts [hunk 3]
```diff
@@ -0,0 +1 @@
+const dummy: any = null;
```
Comment (erikma) [inline types/feathersjs__authentication-jwt/feathersjs__authentication-jwt-tests.ts]: In other contributions I've made it was often a sufficient test to import the main .d.ts and instantiate an instance of related classes. This forces tslint to flag invalid declarations.

File: types/feathersjs__authentication-jwt/index.d.ts [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+// Definitions: https://github.com/feathersjs-ecosystem/feathers-typescript
+// TypeScript Version: 2.2
+
+import { Application } from '@feathersjs/feathers';
+import { Request } from 'express';
+
+export default function feathersAuthenticationJwt(options?: FeathersAuthenticationJWTOptions): () => void;
+
+export interface FeathersAuthenticationJWTOptions {
+    name: string; // the name to use when invoking the authentication Strategy
```
Comment (erikma) [inline types/feathersjs__authentication-jwt/index.d.ts]: Convert line comments to /** doc comments */ to let autocomplete find them

File: types/feathersjs__authentication-local/index.d.ts [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+    Application,
+    Hook,
+    Paginated
+} from '@feathersjs/feathers';
+import { Request } from 'express';
+
+export default function feathersAuthenticationLocal(options?: FeathersAuthenticationLocalOptions): () => void;
+
+export interface FeathersAuthenticationLocalOptions {
+    name: string; // the name to use when invoking the authentication Strategy
```
Comment (erikma) [inline types/feathersjs__authentication-local/index.d.ts]: comments -> doc comments

File: types/feathersjs__errors/index.d.ts [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+    BadGateway: BadGateway;
+    Unavailable: Unavailable;
+}
+
+export function convert(error: any): FeathersError;
+
+export const types: Errors;
+export const errors: Errors;
+
+// todo: (JL) what's with the handlers now?
```
Comment (erikma) [inline types/feathersjs__errors/index.d.ts]: Note: Fixup

File: types/feathersjs__feathers/index.d.ts [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+    remove: Hook | Hook[];
+}
+
+export interface HooksObject {
+    before: Partial<HookMap>;
+    after: Partial<HookMap>;
+    error: Partial<HookMap>;
+}
+
+// todo: figure out what to do: These methods don't actually need to be implemented, so they can be undefined at runtime. Yet making them optional gets cumbersome in strict mode.
```
Comment (erikma) [inline types/feathersjs__feathers/index.d.ts]: Note: Fixup

File: types/feathersjs__rest-client/index.d.ts [hunk 8]
```diff
@@ -0,0 +1,37 @@
+// Type definitions for @feathersjs/rest-client 1.3
+// Project: http://feathersjs.com/
+// Definitions by: Jan Lohage <https://github.com/j2L4e>
+// Definitions: https://github.com/feathersjs-ecosystem/feathers-typescript
+
+// todo: get rid of all the anys
```
Comment (erikma) [inline types/feathersjs__rest-client/index.d.ts]: Note: Fixup

File: types/feathersjs__socket-commons/socket-commons.d.ts [hunk 9]
```diff
@@ -0,0 +1,14 @@
+declare module '@feathersjs/socket-commons' {
+    // (JL) does it have a public API?
```
Comment (erikma) [inline types/feathersjs__socket-commons/socket-commons.d.ts]: Note: Fixup

**Review by erikmav — COMMENTED** [review]
Several TODOs that should be addressed, doc comments to convert, plus consider bulking up test.ts files with real imports of the .d.ts definitions to let tslint have a go.

File: types/feathersjs__feathers/index.d.ts (hunk 7 above)
Comment (j2L4e) [reply types/feathersjs__feathers/index.d.ts]: what's your thought on this?

**Comment by RyanCavanaugh** [comment]
@erikma can you review the new revision? Thanks!

File: types/feathersjs__feathers/index.d.ts (hunk 7 above)
Comment (erikma) [reply types/feathersjs__feathers/index.d.ts]: I would keep them.

**Comment by erikma** [comment]
Thanks for all the doc comments!

**Review by erikmav — APPROVED** [review]

**Comment by eakarpov** [comment]
@j2L4e How should I use the extended Application type with feathersjs__authentication-client?
I've imported the Application type from feathersjs__feathers and it throws compilation errors on logout/authenticate methods.
Thanks

**Comment by j2L4e** [comment]
will be fixed via DefinitelyTyped/DefinitelyTyped#23195

please take any further issues over here: https://github.com/feathersjs-ecosystem/feathers-typescript/issues

**Pull request merged**